File: frontend/lib/models/agent_message.dart
import 'dart:convert';

/// Structured message model for agent-to-agent communication.
/// Follows the standard message envelope defined in the agile methodology docs.
class AgentMessage {
  final int? id;
  final String fromAgent;
  final String toAgent;
  final String messageType; // REQUIREMENTS_PACKAGE, ARCHITECTURE_SPEC, UI_SPEC, etc.
  final int storyId;
  final int projectId;
  final String correlationId;
  final String? payloadJson;
  final bool needsHuman;
  final String? humanQuestionsJson;
  final String status; // PENDING, ROUTED, RESOLVED, ERROR
  final DateTime? createdAt;
  final DateTime? resolvedAt;

  AgentMessage({
    this.id,
    required this.fromAgent,
    required this.toAgent,
    required this.messageType,
    required this.storyId,
    required this.projectId,
    required this.correlationId,
    this.payloadJson,
    this.needsHuman = false,
    this.humanQuestionsJson,
    this.status = 'PENDING',
    this.createdAt,
    this.resolvedAt,
  });

  factory AgentMessage.fromJson(Map<String, dynamic> json) {
    return AgentMessage(
      id: json['id'] as int?,
      fromAgent: json['fromAgent'] as String,
      toAgent: json['toAgent'] as String,
      messageType: json['messageType'] as String,
      storyId: json['storyId'] as int,
      projectId: json['projectId'] as int,
      correlationId: json['correlationId'] as String,
      payloadJson: json['payloadJson'] as String?,
      needsHuman: json['needsHuman'] as bool? ?? false,
      humanQuestionsJson: json['humanQuestionsJson'] as String?,
      status: json['status'] as String? ?? 'PENDING',
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt'] as String) : null,
      resolvedAt: json['resolvedAt'] != null ? DateTime.parse(json['resolvedAt'] as String) : null,
    );
  }

  Map<String, dynamic> toJson() => {
    if (id != null) 'id': id,
    'fromAgent': fromAgent,
    'toAgent': toAgent,
    'messageType': messageType,
    'storyId': storyId,
    'projectId': projectId,
    'correlationId': correlationId,
    if (payloadJson != null) 'payloadJson': payloadJson,
    'needsHuman': needsHuman,
    if (humanQuestionsJson != null) 'humanQuestionsJson': humanQuestionsJson,
    'status': status,
  };

  /// Get human questions as a list (parsed from JSON).
  List<String> get humanQuestions {
    if (humanQuestionsJson == null || humanQuestionsJson!.isEmpty) return [];
    try {
      final List<dynamic> parsed = jsonDecode(humanQuestionsJson!) as List<dynamic>;
      return parsed.map((e) => e.toString()).toList();
    } catch (_) {
      return [humanQuestionsJson!];
    }
  }

  bool get isPendingHuman => needsHuman && status == 'PENDING';
  bool get isRouted => status == 'ROUTED';
  bool get isResolved => status == 'RESOLVED';
  bool get isError => status == 'ERROR';
}

File: backend/src/main/java/com/ucto/backend/entity/AgentMessage.java
package com.ucto.backend.entity;

import jakarta.persistence.*;
import java.time.LocalDateTime;

/**
 * Structured message model for agent-to-agent communication.
 * Follows the standard message envelope defined in docs/message_structure.md.
 * Replaces the old ad-hoc BA Chat message format.
 */
@Entity
@Table(name = "agent_messages")
public class AgentMessage {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String fromAgent;

    @Column(nullable = false)
    private String toAgent;

    @Column(nullable = false)
    private String messageType; // REQUIREMENTS_PACKAGE, ARCHITECTURE_SPEC, UI_SPEC, etc.

    @Column(nullable = false)
    private Long storyId;

    @Column(nullable = false)
    private Long projectId;

    @Column(nullable = false)
    private String correlationId;

    @Column(columnDefinition = "TEXT")
    private String payloadJson; // Structured JSON payload per message_type schema

    @Column(nullable = false)
    private boolean needsHuman;

    @Column(columnDefinition = "TEXT")
    private String humanQuestionsJson; // JSON array of questions for PO

    private LocalDateTime createdAt;
    private LocalDateTime resolvedAt;
    private String status; // PENDING, ROUTED, RESOLVED, ERROR

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        if (status == null) status = "PENDING";
    }

    // Getters and setters
    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }

    public String getFromAgent() { return fromAgent; }
    public void setFromAgent(String fromAgent) { this.fromAgent = fromAgent; }

    public String getToAgent() { return toAgent; }
    public void setToAgent(String toAgent) { this.toAgent = toAgent; }

    public String getMessageType() { return messageType; }
    public void setMessageType(String messageType) { this.messageType = messageType; }

    public Long getStoryId() { return storyId; }
    public void setStoryId(Long storyId) { this.storyId = storyId; }

    public Long getProjectId() { return projectId; }
    public void setProjectId(Long projectId) { this.projectId = projectId; }

    public String getCorrelationId() { return correlationId; }
    public void setCorrelationId(String correlationId) { this.correlationId = correlationId; }

    public String getPayloadJson() { return payloadJson; }
    public void setPayloadJson(String payloadJson) { this.payloadJson = payloadJson; }

    public boolean isNeedsHuman() { return needsHuman; }
    public void setNeedsHuman(boolean needsHuman) { this.needsHuman = needsHuman; }

    public String getHumanQuestionsJson() { return humanQuestionsJson; }
    public void setHumanQuestionsJson(String humanQuestionsJson) { this.humanQuestionsJson = humanQuestionsJson; }

    public LocalDateTime getCreatedAt() { return createdAt; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }

    public LocalDateTime getResolvedAt() { return resolvedAt; }
    public void setResolvedAt(LocalDateTime resolvedAt) { this.resolvedAt = resolvedAt; }

    public String getStatus() { return status; }
    public void setStatus(String status) { this.status = status; }
}